openstack/utils/choose_version.go

```diff
@@ -109,3 +113,94 @@ func ChooseVersion(client *gophercloud.ProviderClient, recognized []*Version) (*
 
 	return highest, endpoint, nil
 }
+
+// GetSupportedMicroversions returns the minimum and maximum microversion that is supported by the ServiceClient Endpoint.
+func GetSupportedMicroversions(client *gophercloud.ServiceClient) (string, string, error) {
+	type valueResp struct {
+		ID         string `json:"id"`
+		Status     string `json:"status"`
+		Version    string `json:"version"`
+		MinVersion string `json:"min_version"`
+	}
+
+	type response struct {
+		Version valueResp `json:"version"`
+	}
+	var resp response
+	_, err := client.Request("GET", client.Endpoint, &gophercloud.RequestOpts{
+		JSONResponse: &resp,
+		OkCodes:      []int{200, 300},
+	})
+
+	if err != nil {
+		return "", "", err
+	}
+
+	return resp.Version.MinVersion, resp.Version.Version, nil
+}
+
+// RequireMicroversion checks that the required microversion is supported and sets it for
+// the ServiceClient.
+func RequireMicroversion(client *gophercloud.ServiceClient, required string) error {
```

Comment:
Would it be possible to change this API to return a copy of the ServiceClient rather than modifying the existing one? The reason for this is because I would like to ensure we always use the same 'base' microversion, and bump it explicitly only for the scope of code sections which require it. This is much simpler to do if you get a copy of the ServiceClient, because you don't have to worry about undoing it under all circumstances when you exist the relevant scope. Something like:

```go
hostnameServiceClient, err := WithMicroversion(serviceClient, "2.90")
if err != nil {
  if IsMicroversionNotSupported(err) {
    ...plan B...
  } else {
    return err
  }
}

// hostnameServiceClient has microversion 2.90
// serviceClient still has its original microversion, so no accidental microversion leaks
```

...or whatever. I'm not so bothered about the actual API so much as the principal of leaving the original unmodified so we don't have to worry about leaks.

Reply:
Done :slightly_smiling_face: